clients/src/main/java/org/apache/kafka/clients/Metadata.java

```diff
@@ -276,27 +347,90 @@ else if (expireMs <= now) {
+            if (!topicsToRetain.test(metadata.topic()))
+                continue;
+
+            if (metadata.error() == Errors.NONE) {
+                if (metadata.isInternal())
+                    internalTopics.add(metadata.topic());
+                for (MetadataResponse.PartitionMetadata partitionMetadata : metadata.partitionMetadata()) {
+                    updatePartitionInfo(metadata.topic(), partitionMetadata, partitionInfo -> {
+                        int epoch = partitionMetadata.leaderEpoch().orElse(RecordBatch.NO_PARTITION_LEADER_EPOCH);
+                        partitions.add(new MetadataCache.PartitionInfoAndEpoch(partitionInfo, epoch));
+                    });
+                }
+            }
+        }
+
+        return new MetadataCache(metadataResponse.clusterId(), new ArrayList<>(metadataResponse.brokers()), partitions,
+                metadataResponse.topicsByError(Errors.TOPIC_AUTHORIZATION_FAILED),
+                metadataResponse.topicsByError(Errors.INVALID_TOPIC_EXCEPTION),
+                internalTopics, metadataResponse.controller());
+    }
+
+    /**
+     * Compute the correct PartitionInfo to cache for a topic+partition and pass to the given consumer.
+     */
+    private void updatePartitionInfo(String topic,
+                                     MetadataResponse.PartitionMetadata partitionMetadata,
+                                     Consumer<PartitionInfo> partitionInfoConsumer) {
+
+        TopicPartition tp = new TopicPartition(topic, partitionMetadata.partition());
+        if (partitionMetadata.leaderEpoch().isPresent()) {
+            int newEpoch = partitionMetadata.leaderEpoch().get();
+            // If the received leader epoch is at least the same as the previous one, update the metadata
+            if (updateLastSeenEpoch(tp, newEpoch, oldEpoch -> newEpoch >= oldEpoch, false)) {
+                partitionInfoConsumer.accept(MetadataResponse.partitionMetaToInfo(topic, partitionMetadata));
+            } else {
+                // Otherwise ignore the new metadata and use the previously cached info
+                PartitionInfo previousInfo = cache.cluster().partition(tp);
+                if (previousInfo != null) {
+                    partitionInfoConsumer.accept(previousInfo);
+                } else {
+                    if (containsTopic(topic)) {
+                        log.debug("Got an older epoch in partition metadata response for {}, but we are not tracking this topic. " +
+                                "Ignoring metadata update for this partition", tp);
+                    } else {
+                        log.warn("Got an older epoch in partition metadata response for {}, but could not find previous partition " +
+                                "info to use. Refusing to update metadata for this partition", tp);
+                    }
+                }
+            }
+        } else {
+            // Old cluster format (no epochs)
```

Comment:
If the metadata API does not support leader epochs, I am considering whether we should clear `lastSeenLeaderEpochs`. Say for example that a user seeks to an offset and provides an epoch. But when we try to fetch metadata for a later epoch, we find the broker is on an older version. This might be possible in a downgrade scenario. In this case, `partitionInfoIfCurrent` would indefinitely return nothing.